### ISSUE TITLE
ERROR: failed to find function RTtranslateMemory, unable to create call

### ISSUE THREAD

**Issue by jjang3** [dyninst/dyninst#1166]
Hello everyone,

I have recently started using Dyninst and have been facing this issue which **originally did not** exist, but now it keeps happening, and I'm sort of out of idea on how to resolve it.  

The stem of this problem arises from this code: `BPatch_dynamicTargetExpr()`. Here is the straightforward code that I have which is causing me this problem:

```
bool instrument_Points(BPatch_binaryEdit* appBin, BPatch_function* instFunc,
                        BPatch_point* instPoints, Dyninst::Address instAddr) 
{
  BPatch_Vector<BPatch_snippet *> openArgs;
  BPatch_snippet* targetAddr = new BPatch_dynamicTargetExpr();
  BPatch_Vector<BPatch_snippet *> targetArgs;
  targetArgs.push_back(targetAddr);
  BPatch_funcCallExpr instFunction(*instFunc, targetArgs);    
  BPatchSnippetHandle *inst_handle =
          appBin->insertSnippet(instFunction, *instPoints, BPatch_callBefore, BPatch_firstSnippet);
}
```

I am patching indirect `jump` or `call` instructions with instrumentation and sending the target address to which these indirect instructions will transfer.

Upon executing the binary, I face this problem, as shown below (I'm using `spack` build of Dyninst):
![Screen Shot 2021-12-05 at 5 27 56 PM](https://user-images.githubusercontent.com/42549923/144766476-44bfc6cb-436c-4c5b-8ce3-8d48525596c4.png)

I could pinpoint the problem because code works as expected if I were to use `openArgs` instead of `targetArgs`.

Environment variables are properly set as shown below:
```
export DYNINST_ROOT=/home/jay/spack/opt/spack/linux-ubuntu18.04-sandybridge/gcc-7.5.0/dy\
ninst-12.0.1-v6g4xox5mxecwwxgwovdjrbq3qyquo7w
export DYNINSTAPI_RT_LIB=$DYNINST_LIBDIR/libdyninstAPI_RT.so
export DYNINST_LIBDIR=$DYNINST_ROOT/lib
export DYNINST_INCDIR=$DYNINST_ROOT/include
export LIBUNWIND_DIR=/home/jay/dyninst-related/libunwind/install
```

Please let me know if I need to provide any additional information or if there is an **alternative** method of doing what I wish to do.

Sincerely,

**Comment by hainest** [comment]
>this issue which originally did not exist, but now it keeps happening, and I'm sort of out of idea on how to resolve it.

Can you detail what changed between when it previously didn't happen and now? Did you change the way you are building the binary you are modifying? Did you change versions of Dyninst?

---
I don't see anything immediately wrong with the code you are showing. Could you do two things:

1. Attach a gdb backtrace (you can drag-and-drop a file to make an attachment)
2. Run your mutator with `DYNINST_DEBUG_INST` set (e.g., `$ DYNINST_DEBUG_INST=1 ./your_mutator`)

Please do these as separate runs of your mutator as the debug logging will make it hard to read the gdb backtrace.

**Comment by jjang3** [comment]
Thank you for your reply @hainest 

1. Can you detail what changed between when it previously didn't happen and now? Did you change the way you are building the binary you are modifying? Did you change versions of Dyninst?

The weird thing is I didn't change anything in regards to Dyninst at all. I simply just relocated `new BPatch_dynamicTargetExpr()` command to somewhere else in the code as a test, and after I compiled that code, I started getting all these issues.

I wasn't sure what I had done wrong. Therefore, I tried using a different machine (installing Dyninst using `spack install dyninst` as before), resulting in the same error, so I'm not sure what is happening since this is recurring even after I did a fresh install.

---

2) I don't see anything immediately wrong with the code you are showing. Could you do two things:

> Attach a gdb backtrace (you can drag-and-drop a file to make an attachment)

> Run your mutator with DYNINST_DEBUG_INST set (e.g., $ DYNINST_DEBUG_INST=1 ./your_mutator)

I have attached the gdb backtrace of the execution. I hope I was able to get the correct information.
[backtrace.txt](https://github.com/dyninst/dyninst/files/7664510/backtrace.txt)

Here is the log that I got after running it with debug flag (which I didn't know was possible before, so this is an excellent thing for me to know)
[dyninstdebug.txt](https://github.com/dyninst/dyninst/files/7664522/dyninstdebug.txt)

Please let me know if any of the information I attached is incorrect.
Kind regards,

**Comment by bigtrak** [comment]
Hello.  I am examining this issue.

Some functionality may have been removed in dyninst-12 which is
required for this to work.  However, there is some question of
whether it worked at all -- which is why that may have been
removed.

If you could try running your tests with the latest dyninst-11 release
to see if they work correctly, it would provide us some feedback on 
those issues -- and ensure that they worked for you previously.

While I am doing so, could you please verify (and correct if needed) an 
inconsistency I noticed while re-reading your issue report?  That will
ensure that there is not a problem at your end causing other issues.

Briefly, examining your script included in the issue which you use 
to setup the DYNINST environment variables, from what you transcribed,
the DYNINST_LIBDIR and DYNINSTAPI_RT_LIB variables are not set correctly, due to a reversed
ordering, where use of dependent variables happens before they are defined.

While it could be a transcription error, best to be certain!

You can verify, before execution, if dyninst can locate it's runtime library via

echo $DYNINSTAPI_RT_LIB
ls -l $DYNINSTAPI_RT_LIB

**Comment by jjang3** [comment]
@bigtrak 

Thank you very much for your reply. I have tried reversing the order of environment variables initialization (didn't realize they are in reversed ordering!) and tried to execute but to no avail, which still raises the same error.

Here is the output of the command you requested:
```
jay@rewriter:~/DynInst$ echo $DYNINSTAPI_RT_LIB
/home/jay/spack/opt/spack/linux-ubuntu18.04-sandybridge/gcc-7.5.0/dyninst-12.0.1-v6g4xox5mxecwwxgwovdjrbq3qyquo7w/lib/libdyninstAPI_RT.so
jay@rewriter:~/DynInst$ ls -l $DYNINSTAPI_RT_LIB
lrwxrwxrwx 1 jay jay 24 Nov 28 19:21 /home/jay/spack/opt/spack/linux-ubuntu18.04-sandybridge/gcc-7.5.0/dyninst-12.0.1-v6g4xox5mxecwwxgwovdjrbq3qyquo7w/lib/libdyninstAPI_RT.so -> libdyninstAPI_RT.so.12.0
jay@rewriter:~/DynInst$ 
```

I haven't got a chance to test it with `dyninst-11`, but let me see if I can try, I have been having a bit of issue with installing `dyninst` from the source, so I relied on `spack` thus far.

Sincerely,

**Comment by hainest** [comment]
@jjang3 We've spotted the problem, and I'm working on a solution.

**Comment by jjang3** [comment]
@hainest Thank you so much for the update! If I need to do anything else to give additional information, please let me know.

**Comment by hainest** [comment]
Apologies. I let this slip off my radar. Here are instructions on building the proposed fix using spack. Let me know how it goes!

    git clone --depth=1 --branch=master https://github.com/dyninst/dyninst.git
    cd dyninst
    wget https://github.com/dyninst/dyninst/pull/1172.patch
    git apply 1172.patch
    spack dev-build --reuse -jN --reuse dyninst@master

You can then find where it put dyninst@master using

    spack find -p dyninst@master

**Comment by hainest** [comment]
@jjang3 Did you have a chance to try out this fix?

**Comment by jjang3** [comment]
@hainest Ah I sincerely apologize for the late response, I have somehow missed your last comment.

I just tried it (as shown below, I have built `dyninst` with your instruction)

![image](https://user-images.githubusercontent.com/42549923/152620509-1382d738-0eda-4ed7-8e90-eaf6b298b042.png), and then I re-compiled my code (without changing anything) and executed it, now it works perfectly fine!

Thank you so much for your help.
